processing/src/main/java/io/druid/segment/filter/JavaScriptFilter.java

```diff
@@ -51,34 +49,57 @@ public ImmutableBitmap getBitmapIndex(final BitmapIndexSelector selector)
   {
     final Context cx = Context.enter();
     try {
-      final Indexed<String> dimValues = selector.getDimensionValues(dimension);
       ImmutableBitmap bitmap;
-      if (dimValues == null) {
-        bitmap = selector.getBitmapFactory().makeEmptyImmutableBitmap();
-      } else {
-        bitmap = selector.getBitmapFactory().union(
-            FunctionalIterable.create(dimValues)
-                              .filter(
-                                  new Predicate<String>()
-                                  {
-                                    @Override
-                                    public boolean apply(@Nullable String input)
-                                    {
-                                      return predicate.applyInContext(cx, input);
-                                    }
-                                  }
-                              )
-                              .transform(
-                                  new com.google.common.base.Function<String, ImmutableBitmap>()
-                                  {
-                                    @Override
-                                    public ImmutableBitmap apply(@Nullable String input)
-                                    {
-                                      return selector.getBitmapIndex(dimension, input);
-                                    }
-                                  }
-                              )
-        );
+
+      boolean hasEmptyDimension = false;
+      Indexed<String> [] dimValuesList = new Indexed[dimension.length];
+      Iterator<String> [] dimValuesIterator = new Iterator[dimension.length];
+      String[] currentDim = new String[dimension.length];
+      for(int idx = 0; idx < dimension.length; idx++) {
+        dimValuesList[idx] = selector.getDimensionValues(dimension[idx]);
+        if (dimValuesList[idx].size() == 0) {
+          hasEmptyDimension = true;
+          break;
+        }
+        dimValuesIterator[idx] = dimValuesList[idx].iterator();
+        if (idx != 0) {
+          currentDim[idx] = dimValuesIterator[idx].next();
+        }
+      }
+
+      bitmap = selector.getBitmapFactory().makeEmptyImmutableBitmap();
+      if (!hasEmptyDimension) {
```

Comment:
@jon-wei As I know, even null value is stored, which means dimValuesList has size of 1 and its only element is empty string (it will be restored as null value at access time).
So I thought no value for any dim means no values in every dims.
Is it right?